statsmodels/tsa/vector_ar/var_model.py

```diff
@@ -1342,39 +1678,188 @@ def test_causality(self, equation, variables, kind='f', signif=0.05,
+
+        caused_ind = [i for i in range(self.neqs) if i not in causing_ind]
+        caused = [self.names[c] for c in caused_ind]
+
+        # Note: JMulTi seems to be using k_ar+1 instead of k_ar
+        k, t, p = self.neqs, self.nobs, self.k_ar
+
+        num_restr = len(causing) * len(caused)  # called N in Lutkepohl
+
+        sigma_u = self.sigma_u
+        vech_sigma_u = util.vech(sigma_u)
+        sig_mask = np.zeros(sigma_u.shape)
+        # set =1 twice to ensure, that all the ones needed are below the main
+        # diagonal:
+        sig_mask[causing_ind, caused_ind] = 1
+        sig_mask[caused_ind, causing_ind] = 1
+        vech_sig_mask = util.vech(sig_mask)
+        inds = np.nonzero(vech_sig_mask)[0]
+
+        # Make restriction matrix
+        C = np.zeros((num_restr, len(vech_sigma_u)), dtype=float)
+        for row in range(num_restr):
+            C[row, inds[row]] = 1
+        Cs = np.dot(C, vech_sigma_u)
+        d = np.linalg.pinv(duplication_matrix(k))
+        Cd = np.dot(C, d)
+        middle = L.inv(chain_dot(Cd, np.kron(sigma_u, sigma_u), Cd.T)) / 2
+
+        wald_statistic = t * chain_dot(Cs.T, middle, Cs)
+        df = num_restr
+        dist = stats.chi2(df)
+
+        pvalue = dist.sf(wald_statistic)
+        crit_value = dist.ppf(1 - signif)
+
+        return CausalityTestResults(causing, caused, wald_statistic,
+                                    crit_value, pvalue, df, signif,
+                                    test="inst", method="wald")
+
+    def test_whiteness_new(self, nlags=10, signif=0.05, adjusted=False):
+        """
+        Test the whiteness of the residuals using the Portmanteau test as
+        described in [1]_, chapter 4.4.3.
+
+        Parameters
+        ----------
+        nlags : int > 0
+        signif : float, between 0 and 1
+        adjusted : bool, default False
 
-        if verbose:
-            summ = output.causality_summary(results, variables, equation, kind)
+        Returns
+        -------
+        results : WhitenessTestResults
 
-            print(summ)
+        References
+        ----------
+        .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+        """
+        def cov(lag):
+            """
+            Parameters
+            ----------
+            lag : int >= 0
+
+            Returns
+            -------
+            result : ndarray (neqs, neqs)
+                The estimated autocovariance matrix of :math:`u_t` for lag
+                `lag`.
+            """
+            u = np.asarray(self.resid).T
+            # the following line is unnecessary in case of OLS-estimation
+            # (mean of residuals =0) but in case other estimation methods are
+            # added we leave here.
+            u -= np.mean(u, axis=1).reshape((u.shape[0], 1))
+            result = np.zeros((self.neqs, self.neqs))
+            for t in range(lag, self.nobs):
+                result += u[:, t:t+1].dot(u[:, t-lag:t-lag+1].T)
```

Comment:
Either I'm not understanding something here or there are many more dot-products being computed than are necessary.  Aside from de-meaning the residuals (which could be done once, outside the loop), couldn't we just call `_compute_acov`?